wagtail/wagtail
# Consolidate all calls to unidecode & remove unidecode from form field clean_name

This PR covers a few items that are interrelated, I am happy to split into two PRs if that is more suitable.

The underlying goal is to centralise the usage of `unidecode` in preparation for a future removal of reliance on this package due to licensing issues - see #3311

## Overview
* *Builds on the work by @ashiazed & @mvantellingen in separate PRs*
* Moves `unidecode` to a central wagtail.core.utils function so it can be replaced/removed easier in the future, superseeds the work done on #5508
* Provides a safe migration path for existing applications with Wagtail Forms where saved data must still honour the `unidecode` kebab-case clean name on fields.
* Wagtail forms will now generate a clean name on fields as a field, not a method, this provides the ability to do the migration path above but has been done in a way that no longer relies on `unidecode`. As per https://github.com/wagtail/wagtail/pull/3088#issuecomment-284877264
* Finally, in the change for the clean name on form fields, we have adopted `snake_case` instead of `kebab-case`, this makes these easier to use in templates (for simple cases), superseeds #3088
* Small fixes; add some basic tests for the camel case conversion, add tests for the new wrapper function around `unidecode`, refine duplicate field labels to compare on their eventual clean_name instead of just a basic string comparison.

## How to Validate this PR
* First, ensure you have an app running locally that has at least one form page, ideally two.* Add some new fields to each form, try to use some unicode values inside the labels - https://www.ltg.ed.ac.uk/~richard/unicode-sample.html
* Submit some responses on each of those forms
* Check out the PR on the local app, first run manage.py runserver, you should receive a warning that there are migrations to run AND the application should not crash.
* Run makemigrations, the existing form field model should be updated with an addied field `clean_name`, then migrate. When running these, it should not crash.
* Finally, load up the server, this is where the checks should run and because the fields are added you should get an info message advising that some clean_name updates have happened.* Check that you can still see the entries saved as form responses previously.
* Attempt to create a duplicate field in an existing page with some caps/non-caps variation of an existing field.

## Notes
* The new clean names may not be as visually nice for complex unicode scenarios, however these are used internally only and only those doing customisations in rendering will get visibility into these, even when exporting to csv the actual user entered label is resolved.
* We can refine the `safe_snake_case` function in the future knowing it will not break existing data stored.
* This PR does not handle scenarios where a field is added to a form with name 'A' and then changed to name 'B' and updating the clean_name from there, this is intentional so that data is not lost when labels change. Plus implementing this would open us up to a lot of complexity.
* There are massive changes to the form tests and test data, however I have intentionally set one model in the test data to have clean_name blank. This is to ensure we can test that the checks run and will update data with the legacy clean name.
* This PR does not remove the requirement of unidecode, but paves the way for this to be done in the next release as there is now a clear migration strategy for existing applications.

## Checks
* Do the tests still pass? 👍 
* Does the code comply with the style guide? 👍 
* For Python changes: Have you added tests to cover the new/fixed behaviour? 👍 
* For new features: Has the documentation been updated accordingly? N/A - we may want to add some upgrade considerations but the intent of this change is that it happens seamlessly.
